pkg/keyspace/tso_keyspace_group.go

```diff
@@ -72,42 +122,137 @@ func (m *GroupManager) GetKeyspaceGroupByID(id uint32) (*endpoint.KeyspaceGroup,
+	m.groups[userKind].Remove(id)
+
 	return nil
 }
 
-func (m *GroupManager) saveKeyspaceGroups(keyspaceGroups []*endpoint.KeyspaceGroup) error {
+func (m *GroupManager) saveKeyspaceGroups(keyspaceGroups []*endpoint.KeyspaceGroup, overwrite bool) error {
 	return m.store.RunInTxn(m.ctx, func(txn kv.Txn) error {
 		for _, keyspaceGroup := range keyspaceGroups {
 			// TODO: add replica count
 			newKG := &endpoint.KeyspaceGroup{
-				ID:       keyspaceGroup.ID,
-				UserKind: keyspaceGroup.UserKind,
+				ID:        keyspaceGroup.ID,
+				UserKind:  keyspaceGroup.UserKind,
+				Keyspaces: keyspaceGroup.Keyspaces,
 			}
 			// Check if keyspace group has already existed.
 			oldKG, err := m.store.LoadKeyspaceGroup(txn, keyspaceGroup.ID)
 			if err != nil {
 				return err
 			}
-			if oldKG != nil {
+			if oldKG != nil && !overwrite {
 				return ErrKeyspaceGroupExists
 			}
 			m.store.SaveKeyspaceGroup(txn, newKG)
 		}
 		return nil
 	})
 }
+
+// GetAvailableKeyspaceGroupIDByKind returns the available keyspace group id by user kind.
+func (m *GroupManager) GetAvailableKeyspaceGroupIDByKind(userKind endpoint.UserKind) (string, error) {
+	m.RLock()
+	defer m.RUnlock()
+	groups, ok := m.groups[userKind]
+	if !ok {
+		return "", errors.Errorf("user kind %s not found", userKind)
+	}
+	kg := groups.Top()
+	return strconv.FormatUint(uint64(kg.ID), 10), nil
+}
+
+// UpdateKeyspaceForGroup updates the keyspace field for the keyspace group.
+func (m *GroupManager) UpdateKeyspaceForGroup(userKind endpoint.UserKind, groupID string, keyspaceID uint32, mutation string) error {
+	id, err := strconv.ParseUint(groupID, 10, 64)
+	if err != nil {
+		return err
+	}
+
+	m.Lock()
+	defer m.Unlock()
+	kg := m.groups[userKind].Get(uint32(id))
+	switch mutation {
+	case opAdd:
+		if !slice.Contains(kg.Keyspaces, keyspaceID) {
+			kg.Keyspaces = append(kg.Keyspaces, keyspaceID)
+			m.groups[userKind].Put(kg)
+		}
+	case opDelete:
+		if slice.Contains(kg.Keyspaces, keyspaceID) {
+			kg.Keyspaces = slice.Remove(kg.Keyspaces, keyspaceID)
+			m.groups[userKind].Put(kg)
+		}
+	}
+
+	return m.saveKeyspaceGroups([]*endpoint.KeyspaceGroup{kg}, true)
+}
+
+// UpdateKeyspaceGroup updates the keyspace group.
+func (m *GroupManager) UpdateKeyspaceGroup(oldGroupID, newGroupID string, oldUserKind, newUserKind endpoint.UserKind, keyspaceID uint32) error {
+	oldID, err := strconv.ParseUint(oldGroupID, 10, 64)
+	if err != nil {
+		return err
+	}
+	newID, err := strconv.ParseUint(newGroupID, 10, 64)
+	if err != nil {
+		return err
+	}
+
+	m.Lock()
+	defer m.Unlock()
+	oldKG := m.groups[oldUserKind].Get(uint32(oldID))
+	if oldKG == nil {
+		return errors.Errorf("keyspace group %s not found in %s group", oldGroupID, oldUserKind)
+	}
+	newKG := m.groups[newUserKind].Get(uint32(newID))
+	if newKG == nil {
+		return errors.Errorf("keyspace group %s not found in %s group", newGroupID, newUserKind)
+	}
+
+	if !slice.Contains(newKG.Keyspaces, keyspaceID) {
```

Comment:
O(n). Another reason we might need a more efficient and compact data structure like tree map

Reply:
ditto